{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0a6ae9c",
   "metadata": {},
   "source": [
    "# Safi_10859_250321B9_se - atac_seq\n",
    "This notebook will create all the necessary files, scripts and folders to pre-process the aforementioned project. Is designed to be used in a jupyter server deployed in a system running SLURM. The majority of the scripts and heavy-lifting processes are wrapped up in sbatch scripts.As an end user, in order to pre-process your samples provided in the spread sheet, you will simply need to *run the entire notebook* (Cell > Run all) and the system should take care of the rest for you.\n",
    "#### Create necessary folder(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d87bfbbd",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /work/aeb84/data/atac_seq/metadata\n",
    "mkdir -p /work/aeb84/data/atac_seq/raw_reads\n",
    "mkdir -p /work/aeb84/data/atac_seq/processed_raw_reads\n",
    "mkdir -p /work/aeb84/processing/atac_seq/scripts\n",
    "mkdir -p /work/aeb84/processing/atac_seq/jsons\n",
    "mkdir -p /work/aeb84/processing/atac_seq/logs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb3d67ed",
   "metadata": {},
   "source": [
    "Save metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c93fd8a1",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /work/aeb84/data/atac_seq/metadata/atac_seq_download_metadata.Safi_10859_250321B9_se.txt\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/aeb84/data/atac_seq/metadata/atac_seq_download_metadata.Safi_10859_250321B9_se.txt\n",
    "Sequencing core user\tSequencing core project\tSequencing core password\tSequencing core library name\tName\tPaired-end or single-end\tGenome\tLibrary type\tUmis\tControl\tStrand specificity\tBlacklist removal\tWith ercc spike-in\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_18hVer_1\tHFF.1KPa.18hVer.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_18hVer_3\tHFF.1KPa.18hVer.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_18hVer_2\tHFF.1KPa.18hVer.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_4hVer_1\tHFF.1KPa.4hVer.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_4hVer_2\tHFF.1KPa.4hVer.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_4hVer_3\tHFF.1KPa.4hVer.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_CCG_2\tHFF.1KPa.CCG.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_CCG_3\tHFF.1KPa.CCG.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_DMSO_1\tHFF.1KPa.DMSO.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_DMSO_2\tHFF.1KPa.DMSO.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_1KPa_DMSO_3\tHFF.1KPa.DMSO.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_18hVer_1\tHFF.50KPa.18hVer.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_18hVer_2\tHFF.50KPa.18hVer.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_4hVer_1\tHFF.50KPa.4hVer.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_18hVer_3\tHFF.50KPa.18hVer.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_CCG_1\tHFF.50KPa.CCG.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_4hVer_2\tHFF.50KPa.4hVer.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_4hVer_3\tHFF.50KPa.4hVer.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_CCG_2\tHFF.50KPa.CCG.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_CCG_3\tHFF.50KPa.CCG.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_DMSO_1\tHFF.50KPa.DMSO.rep1\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_DMSO_2\tHFF.50KPa.DMSO.rep2\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n",
    "safi_10859\tSafi_10859_250321B9\tzkOC7qCadKdy\tHFF_50KPa_DMSO_3\tHFF.50KPa.DMSO.rep3\tSE\thg38\tATAC-seq\t\t\t\tTrue\t\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb42374c",
   "metadata": {},
   "source": [
    "### FASTQ files already available locally!!\n",
    "Please, make sure the FASTQ files are correctly named, decompressed and located/symlinked in:\n",
    "\n",
    "**\n",
    "/work/aeb84/data/atac_seq/processed_raw_reads/Safi_10859_250321B9_se\n",
    "**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ce81a1f7-af66-4dde-8f48-670f509675d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /work/aeb84/data/atac_seq/processed_raw_reads/Safi_10859_250321B9_se/\n",
    "cd /work/aeb84/data/atac_seq/processed_raw_reads/Safi_10859_250321B9_se/\n",
    "ln -s /work/aeb84/data/atac_seq/processed_raw_reads/Safi_10859_250321B9/*R1*fastq.gz ./\n",
    "rename .R1.fastq .fastq *.R1.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "236da998",
   "metadata": {},
   "source": [
    "#### Create JSON files for CWL pipeline files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3f29f726",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /work/aeb84/processing/atac_seq/scripts/cwl_json_gen_Safi_10859_250321B9_se.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/aeb84/processing/atac_seq/scripts/cwl_json_gen_Safi_10859_250321B9_se.sh\n",
    "#!/bin/bash\n",
    "set -euo pipefail\n",
    "\n",
    "ORDER=Safi_10859_250321B9_se\n",
    "PROCESSED_DATA_DIR=/work/aeb84/data/atac_seq/processed_raw_reads/${ORDER}\n",
    "METADATA=/work/aeb84/data/atac_seq/metadata/atac_seq_download_metadata.Safi_10859_250321B9_se.txt\n",
    "\n",
    "python /hpc/group/gersbachlab/software/IGVF-cwl/v1.0/json-generator/run.py \\\n",
    "    -m ${METADATA} \\\n",
    "    -d ${PROCESSED_DATA_DIR} \\\n",
    "    -o /work/aeb84/processing/atac_seq/jsons \\\n",
    "    -t atac-seq \\\n",
    "    --fastq-gzipped \\\n",
    "    --mem 24000 \\\n",
    "    --nthreads 16 \\\n",
    "    --separate-jsons\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74167d75",
   "metadata": {},
   "source": [
    "Execute file to create JSON files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b397c409",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "%%script --out blocking_job_str bash\n",
    "source /hpc/group/gersbachlab/software/miniforge3/bin/activate /hpc/group/gersbachlab/aeb84/mamba_envs/cwl10\n",
    "sbatch \\\n",
    " -o /work/aeb84/processing/atac_seq/logs/Safi_10859_250321B9_se_cwl_json_gen.out \\\n",
    " -p common \\\n",
    " /work/aeb84/processing/atac_seq/scripts/cwl_json_gen_Safi_10859_250321B9_se.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "eac389cc-563a-498c-9765-138beccaab07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load /work/aeb84/processing/atac_seq/jsons/atac_seq_download_metadata.Safi_10859_250321B9_se-se-blacklist-removal-6.json\n",
    "{\n",
    "    \"input_fastq_files\": [\n",
    "        { \"class\": \"File\", \"path\": \"/work/aeb84/data/atac_seq/processed_raw_reads/Safi_10859_250321B9_se/HFF.1KPa.CCG.rep2.fastq.gz\" }\n",
    "    ],\n",
    "    \"as_narrowPeak_file\": { \"class\": \"File\", \"path\": \"/hpc/group/gersbachlab/Reference_Data/ENCODE/kent/src/hg/lib/encode/narrowPeak.as\" },\n",
    "    \"genome_effective_size\": \"hs\",\n",
    "    \"default_adapters_file\": { \"class\": \"File\", \"path\": \"/hpc/group/gersbachlab/Reference_Data/Adapters/default_adapters.fasta\" },\n",
    "    \"genome_ref_first_index_file\": { \"class\": \"File\", \"path\": \"/hpc/group/gersbachlab/Reference_Data/IGVF/GRCh38/bowtie_GRCh38/IGVFFI0653VCGH_GRCh38.1.ebwt\"},\n",
    "    \"genome_sizes_file\": { \"class\": \"File\", \"path\": \"/hpc/group/gersbachlab/Reference_Data/IGVF/GRCh38/GCA_000001405.15_GRCh38_no_alt_analysis_set.sizes\" },\n",
    "    \"ENCODE_blacklist_bedfile\": { \"class\": \"File\", \"path\": \"/hpc/group/gersbachlab/Reference_Data/IGVF/GRCh38/hg38.blacklist.v3.bed\" },\n",
    "    \"nthreads_qc\": 16,\n",
    "    \"nthreads_trimm\": 16,\n",
    "    \"nthreads_map\": 16,\n",
    "    \"nthreads_peakcall\": 16,\n",
    "    \"nthreads_quant\": 16,\n",
    "    \"trimmomatic_jar_path\": \"/opt/apps/rhel7/Trimmomatic-0.39/trimmomatic-0.39.jar\",\n",
    "    \"trimmomatic_java_opts\": \"-Xms12000m -Xmx24000m\",\n",
    "    \"picard_jar_path\": \"/opt/apps/rhel7/picard/2.18.2/picard.jar\",\n",
    "    \"picard_java_opts\": \"-Xms12000m -Xmx24000m\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "432ee91d-429d-4ded-a054-4d80dbff3510",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/hpc/group/gersbachlab/software/IGVF-cwl/v1.0/json-generator/run.py\", line 433, in <module>\n",
      "    main()\n",
      "  File \"/hpc/group/gersbachlab/software/IGVF-cwl/v1.0/json-generator/run.py\", line 426, in main\n",
      "    for json_str, conf_name, idx in meta_parser.parse_metadata(args.data_dir.rstrip('/')):\n",
      "  File \"/hpc/group/gersbachlab/software/IGVF-cwl/v1.0/json-generator/run.py\", line 186, in parse_metadata\n",
      "    for si, s in enumerate(sorted(samples_genomes)):\n",
      "TypeError: '<' not supported between instances of 'dict' and 'dict'\n"
     ]
    }
   ],
   "source": [
    "!cat /work/aeb84/processing/atac_seq/logs/Safi_10859_250321B9_se_cwl_json_gen.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9dbdbe41-abdf-4f20-b609-a6db25267725",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequencing core user</th>\n",
       "      <th>Sequencing core project</th>\n",
       "      <th>Sequencing core password</th>\n",
       "      <th>Sequencing core library name</th>\n",
       "      <th>Name</th>\n",
       "      <th>Paired-end or single-end</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Library type</th>\n",
       "      <th>Umis</th>\n",
       "      <th>Control</th>\n",
       "      <th>Strand specificity</th>\n",
       "      <th>Blacklist removal</th>\n",
       "      <th>With ercc spike-in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_18hVer_1</td>\n",
       "      <td>HFF.1KPa.18hVer.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_18hVer_3</td>\n",
       "      <td>HFF.1KPa.18hVer.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_18hVer_2</td>\n",
       "      <td>HFF.1KPa.18hVer.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_4hVer_1</td>\n",
       "      <td>HFF.1KPa.4hVer.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_4hVer_2</td>\n",
       "      <td>HFF.1KPa.4hVer.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_4hVer_3</td>\n",
       "      <td>HFF.1KPa.4hVer.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_CCG_2</td>\n",
       "      <td>HFF.1KPa.CCG.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_CCG_3</td>\n",
       "      <td>HFF.1KPa.CCG.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_DMSO_1</td>\n",
       "      <td>HFF.1KPa.DMSO.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_DMSO_2</td>\n",
       "      <td>HFF.1KPa.DMSO.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_1KPa_DMSO_3</td>\n",
       "      <td>HFF.1KPa.DMSO.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_18hVer_1</td>\n",
       "      <td>HFF.50KPa.18hVer.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_18hVer_2</td>\n",
       "      <td>HFF.50KPa.18hVer.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_4hVer_1</td>\n",
       "      <td>HFF.50KPa.4hVer.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_18hVer_3</td>\n",
       "      <td>HFF.50KPa.18hVer.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_CCG_1</td>\n",
       "      <td>HFF.50KPa.CCG.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_4hVer_2</td>\n",
       "      <td>HFF.50KPa.4hVer.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_4hVer_3</td>\n",
       "      <td>HFF.50KPa.4hVer.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_CCG_2</td>\n",
       "      <td>HFF.50KPa.CCG.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_CCG_3</td>\n",
       "      <td>HFF.50KPa.CCG.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_DMSO_1</td>\n",
       "      <td>HFF.50KPa.DMSO.rep1</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_DMSO_2</td>\n",
       "      <td>HFF.50KPa.DMSO.rep2</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>safi_10859</td>\n",
       "      <td>Safi_10859_250321B9</td>\n",
       "      <td>zkOC7qCadKdy</td>\n",
       "      <td>HFF_50KPa_DMSO_3</td>\n",
       "      <td>HFF.50KPa.DMSO.rep3</td>\n",
       "      <td>SE</td>\n",
       "      <td>hg38</td>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sequencing core user Sequencing core project Sequencing core password  \\\n",
       "0            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "1            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "2            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "3            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "4            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "5            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "6            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "7            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "8            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "9            safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "10           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "11           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "12           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "13           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "14           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "15           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "16           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "17           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "18           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "19           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "20           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "21           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "22           safi_10859     Safi_10859_250321B9             zkOC7qCadKdy   \n",
       "\n",
       "   Sequencing core library name                   Name  \\\n",
       "0             HFF_1KPa_18hVer_1   HFF.1KPa.18hVer.rep1   \n",
       "1             HFF_1KPa_18hVer_3   HFF.1KPa.18hVer.rep3   \n",
       "2             HFF_1KPa_18hVer_2   HFF.1KPa.18hVer.rep2   \n",
       "3              HFF_1KPa_4hVer_1    HFF.1KPa.4hVer.rep1   \n",
       "4              HFF_1KPa_4hVer_2    HFF.1KPa.4hVer.rep2   \n",
       "5              HFF_1KPa_4hVer_3    HFF.1KPa.4hVer.rep3   \n",
       "6                HFF_1KPa_CCG_2      HFF.1KPa.CCG.rep2   \n",
       "7                HFF_1KPa_CCG_3      HFF.1KPa.CCG.rep3   \n",
       "8               HFF_1KPa_DMSO_1     HFF.1KPa.DMSO.rep1   \n",
       "9               HFF_1KPa_DMSO_2     HFF.1KPa.DMSO.rep2   \n",
       "10              HFF_1KPa_DMSO_3     HFF.1KPa.DMSO.rep3   \n",
       "11           HFF_50KPa_18hVer_1  HFF.50KPa.18hVer.rep1   \n",
       "12           HFF_50KPa_18hVer_2  HFF.50KPa.18hVer.rep2   \n",
       "13            HFF_50KPa_4hVer_1   HFF.50KPa.4hVer.rep1   \n",
       "14           HFF_50KPa_18hVer_3  HFF.50KPa.18hVer.rep3   \n",
       "15              HFF_50KPa_CCG_1     HFF.50KPa.CCG.rep1   \n",
       "16            HFF_50KPa_4hVer_2   HFF.50KPa.4hVer.rep2   \n",
       "17            HFF_50KPa_4hVer_3   HFF.50KPa.4hVer.rep3   \n",
       "18              HFF_50KPa_CCG_2     HFF.50KPa.CCG.rep2   \n",
       "19              HFF_50KPa_CCG_3     HFF.50KPa.CCG.rep3   \n",
       "20             HFF_50KPa_DMSO_1    HFF.50KPa.DMSO.rep1   \n",
       "21             HFF_50KPa_DMSO_2    HFF.50KPa.DMSO.rep2   \n",
       "22             HFF_50KPa_DMSO_3    HFF.50KPa.DMSO.rep3   \n",
       "\n",
       "   Paired-end or single-end Genome Library type  Umis  Control  \\\n",
       "0                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "1                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "2                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "3                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "4                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "5                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "6                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "7                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "8                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "9                        SE   hg38     ATAC-seq   NaN      NaN   \n",
       "10                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "11                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "12                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "13                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "14                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "15                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "16                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "17                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "18                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "19                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "20                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "21                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "22                       SE   hg38     ATAC-seq   NaN      NaN   \n",
       "\n",
       "    Strand specificity  Blacklist removal  With ercc spike-in  \n",
       "0                  NaN               True                 NaN  \n",
       "1                  NaN               True                 NaN  \n",
       "2                  NaN               True                 NaN  \n",
       "3                  NaN               True                 NaN  \n",
       "4                  NaN               True                 NaN  \n",
       "5                  NaN               True                 NaN  \n",
       "6                  NaN               True                 NaN  \n",
       "7                  NaN               True                 NaN  \n",
       "8                  NaN               True                 NaN  \n",
       "9                  NaN               True                 NaN  \n",
       "10                 NaN               True                 NaN  \n",
       "11                 NaN               True                 NaN  \n",
       "12                 NaN               True                 NaN  \n",
       "13                 NaN               True                 NaN  \n",
       "14                 NaN               True                 NaN  \n",
       "15                 NaN               True                 NaN  \n",
       "16                 NaN               True                 NaN  \n",
       "17                 NaN               True                 NaN  \n",
       "18                 NaN               True                 NaN  \n",
       "19                 NaN               True                 NaN  \n",
       "20                 NaN               True                 NaN  \n",
       "21                 NaN               True                 NaN  \n",
       "22                 NaN               True                 NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.read_csv('/work/aeb84/data/atac_seq/metadata/atac_seq_download_metadata.Safi_10859_250321B9_se.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a242e04c",
   "metadata": {},
   "source": [
    "Extract blocking job id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8ed89fe8",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "import re\n",
    "blocking_job = re.match('Submitted batch job (\\d+).*', blocking_job_str).group(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "826223bd",
   "metadata": {},
   "source": [
    "#### Create SLURM array master bash file for se-blacklist-removal samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2c3b9e16",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /work/aeb84/processing/atac_seq/scripts/Safi_10859_250321B9_se-se-blacklist-removal.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/aeb84/processing/atac_seq/scripts/Safi_10859_250321B9_se-se-blacklist-removal.sh\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=cwl_atac_seq\n",
    "#SBATCH --output=/work/aeb84/processing/atac_seq/logs/Safi_10859_250321B9_se-se-blacklist-removal-%a.out\n",
    "#SBATCH --mail-user=aeb84@duke.edu\n",
    "#SBATCH --mail-type=FAIL,END\n",
    "#SBATCH --mem=24000\n",
    "#SBATCH --cpus-per-task=16\n",
    "\n",
    "set -euo pipefail\n",
    "\n",
    "# export PATH=\"/hpc/group/gersbachlab/software/bin:$PATH\"\n",
    "export PATH=\"/hpc/group/gersbachlab/software/ucsc:$PATH\"\n",
    "export PATH=\"/hpc/group/gersbachlab/software/cwl/bin:$PATH\"\n",
    "# export PATH=\"/hpc/group/gersbachlab/software/rsem-1.2.21/:$PATH\"\n",
    "export PATH=\"/hpc/group/gersbachlab/software/phantompeakqualtools-1.2/:$PATH\"\n",
    "# export PATH=\"/hpc/group/gersbachlab/software/miniforge3/envs//hpc/group/gersbachlab/aeb84/mamba_envs/cwl10/bin:$PATH\"\n",
    "\n",
    "module load Bedtools/2.30.0\n",
    "module load FastQC/0.11.7\n",
    "module load Bowtie/1.2.2\n",
    "module load Bowtie2\n",
    "## module load samtools/1.14-rhel8. # samtools v1.18 already install in cwl10\n",
    "## module load Java/1.8.0_60\n",
    "module load Picard/2.18.2\n",
    "module load Trimmomatic/0.39\n",
    "module load Boost/1.75-rhel8\n",
    "\n",
    "\n",
    "# For Fastqc\n",
    "export DISPLAY=:0.0\n",
    "\n",
    "# Make sure temporary files and folders are created in a specific folder\n",
    "mkdir -p /work/aeb84/processing/atac_seq/tmpdirs/tmp-Safi_10859_250321B9_se-se-blacklist-removal-${SLURM_ARRAY_TASK_ID}-\n",
    "export TMPDIR=\"/work/aeb84/processing/atac_seq/tmpdirs/tmp-Safi_10859_250321B9_se-se-blacklist-removal-${SLURM_ARRAY_TASK_ID}-\"\n",
    "\n",
    "cwltool --debug \\\n",
    "    --non-strict \\\n",
    "    --preserve-environment PATH \\\n",
    "    --preserve-environment DISPLAY \\\n",
    "    --preserve-environment TMPDIR \\\n",
    "    --cachedir ${TMPDIR} \\\n",
    "    --outdir /work/aeb84/processing/atac_seq/Safi_10859_250321B9_se-se-blacklist-removal  \\\n",
    "    --no-container \\\n",
    "    /hpc/group/gersbachlab/software/IGVF-cwl/v1.0/ATAC-seq_pipeline/pipeline-se-blacklist-removal.cwl \\\n",
    "    /work/aeb84/processing/atac_seq/jsons/atac_seq_download_metadata.Safi_10859_250321B9_se-se-blacklist-removal-${SLURM_ARRAY_TASK_ID}.json\n",
    "\n",
    "# Delete any tmpdir not removed by cwltool\n",
    "# rm -rf /work/aeb84/processing/atac_seq/tmpdirs/tmp-Safi_10859_250321B9_se-se-blacklist-removal-${SLURM_ARRAY_TASK_ID}-\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd407321",
   "metadata": {},
   "source": [
    "Execute SLURM array master file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5d0130ae",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "%%script --out blocking_job_str bash -s \"$blocking_job\"\n",
    "source /hpc/group/gersbachlab/software/miniforge3/bin/activate /hpc/group/gersbachlab/aeb84/mamba_envs/cwl10\n",
    "sbatch \\\n",
    " -p common \\\n",
    " --array 0,3,4,11,15,16,22 \\\n",
    " /work/aeb84/processing/atac_seq/scripts/Safi_10859_250321B9_se-se-blacklist-removal.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9153291e-ca22-443c-adf2-5376f5229e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /hpc/group/gersbachlab/aeb84/reddylab_utils/scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7b9ff5a6-8b3a-4ce3-b87e-1ef0b393fa80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /hpc/group/gersbachlab/aeb84/reddylab_utils/scripts/ataqc.make_tss_plot.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /hpc/group/gersbachlab/aeb84/reddylab_utils/scripts/ataqc.make_tss_plot.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "from os import path\n",
    "import argparse\n",
    "\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('--tss', help=\"TSS reference file\")\n",
    "    parser.add_argument('--bam', help=\"BAM file\")\n",
    "    parser.add_argument('--read-length', type=int, default=50, help=\"Read length\")\n",
    "    parser.add_argument('--processes', type=int, default=1, help=\"Number of threads (cpus in SLURM) to use.\")\n",
    "    parser.add_argument('--outfile-prefix', help=\"Rootname to output TSS plot\")\n",
    "    parser.add_argument('--chromosome-sizes', help=\"Chromosome sizes file\")\n",
    "    parser.add_argument('--ataqc-rootpath', help=\"Root path to ATAQC package\", default=\"/data/reddylab/software\")\n",
    "\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    # Load ATAQC functions\n",
    "    path.sys.path.append(args.ataqc_rootpath)\n",
    "    from ataqc import run_ataqc\n",
    "\n",
    "    plotfile, plotfile_large, tss_enrich_point_val = \\\n",
    "        run_ataqc.make_tss_plot(bam_file=args.bam,\n",
    "                                tss=args.tss,\n",
    "                                prefix=args.outfile_prefix,\n",
    "                                chromsizes=args.chromosome_sizes,\n",
    "                                read_len=args.read_length,\n",
    "                                processes=args.processes)\n",
    "\n",
    "    print(\"Done!\")\n",
    "    print(\"Save files:\")\n",
    "    print(\"\\t%s\\n\\t%s\" % (plotfile, plotfile_large))\n",
    "    print(\"%s TSS enrichemnt point value:\" % args.bam, tss_enrich_point_val)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fb8568f7-ecd7-47ff-b9a8-39b79571541f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bedtools 2.30.0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 27724659\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# module load bedtools2\n",
    "module load Bedtools/2.30.0\n",
    "\n",
    "source /hpc/group/gersbachlab/software/miniforge3/bin/activate /hpc/group/gersbachlab/aeb84/mamba_envs/alex_py2\n",
    "BAMFILES=($(/bin/ls -1 /work/aeb84/processing/atac_seq/Safi_10859_250321B9_se-se-blacklist-removal/H*masked.dups_marked.bam))\n",
    "#     --array=1-$((${#BAMFILES[@]}-1)) \\\n",
    "sbatch \\\n",
    "    -p common \\\n",
    "    --cpus-per-task 4 \\\n",
    "    --array=1-$((${#BAMFILES[@]}-1)) \\\n",
    "    -o /work/aeb84/processing/atac_seq/logs/ataqc.Safi_10859_250321B9_se-se-blacklist-removal.%a.out \\\n",
    "    <<'EOF'\n",
    "#!/bin/bash\n",
    "BAMFILES=($(/bin/ls -1 /work/aeb84/processing/atac_seq/Safi_10859_250321B9_se-se-blacklist-removal/H*masked.dups_marked.bam))\n",
    "BAM=${BAMFILES[$SLURM_ARRAY_TASK_ID]}\n",
    "# \n",
    "python2 /hpc/group/gersbachlab/aeb84/reddylab_utils/scripts/ataqc.make_tss_plot.py \\\n",
    "    --bam ${BAM} \\\n",
    "    --tss /hpc/group/gersbachlab/Reference_Data/Refseq/GRCh38/GRCh38_latest_genomic.ucsc_chromosomes.promoters.protein_coding.no_alt.bed \\\n",
    "    --read-length 100 \\\n",
    "    --processes 4 \\\n",
    "    --outfile-prefix ${BAM/.bam/.ataqc} \\\n",
    "    --ataqc-rootpath /hpc/group/gersbachlab/aeb84/software \\\n",
    "    --chromosome-sizes \\\n",
    "        /hpc/group/gersbachlab/Reference_Data/Genomes/hg38/GCA_000001405.15_GRCh38_no_alt_analysis_set.sizes\n",
    "EOF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "eb02f069-5e8a-4fd0-9010-99065cbfda5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Python 2.7.15\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source /hpc/group/gersbachlab/software/miniforge3/bin/activate base\n",
    "source /hpc/group/gersbachlab/software/miniforge3/bin/activate /hpc/group/gersbachlab/aeb84/mamba_envs/alex_py2\n",
    "python2 --version\n",
    "python2 -c \"import pysam\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28783e4e",
   "metadata": {},
   "source": [
    "Extract blocking job id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c77df6a0",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "import re\n",
    "blocking_job = re.match('Submitted batch job (\\d+).*', blocking_job_str).group(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beac144b",
   "metadata": {},
   "source": [
    "#### Create QC generating script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "55b62012",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /work/aeb84/processing/atac_seq/scripts/generate_qc_cell_Safi_10859_250321B9_se-se-blacklist-removal.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile /work/aeb84/processing/atac_seq/scripts/generate_qc_cell_Safi_10859_250321B9_se-se-blacklist-removal.sh\n",
    "#!/bin/bash\n",
    "set -euo pipefail\n",
    "\n",
    "#SBATCH --job-name=qc\n",
    "#SBATCH --output=/work/aeb84/processing/atac_seq/logs/qc_gen.Safi_10859_250321B9_se-se-blacklist-removal.out\n",
    "\n",
    "source /hpc/group/gersbachlab/software/miniforge3/bin/activate /hpc/group/gersbachlab/aeb84/mamba_envs/cwl10\n",
    "cd /work/aeb84/processing/atac_seq/Safi_10859_250321B9_se-se-blacklist-removal\n",
    "\n",
    "python /hpc/group/gersbachlab/software/cwl/bin/generate_stats_atacseq_single_end.py ./ \\\n",
    "    -samples $(/bin/ls -1 *PBC.txt | sed 's@.trimmed.PBC.txt@@') \\\n",
    "> qc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "439ec5aa-eb38-4bdc-981a-02a9414e23a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /hpc/group/gersbachlab/aeb84/software/cwl/generate_stats_atacseq_single_end.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /hpc/group/gersbachlab/aeb84/software/cwl/generate_stats_atacseq_single_end.py\n",
    "#!/usr/bin/env python\n",
    "from collections import defaultdict as dd\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description='Generate QC stats for the ChIP-seq CWL pipeline')\n",
    "parser.add_argument('outdir', type=str, help='Directory with CWL ChIP-seq pipeline output files')\n",
    "parser.add_argument('-samples', required=True, nargs='+', help='Sample names for which stats are computed')\n",
    "parser.add_argument('--nonmasked', required=False, action='store_true', help='If no blacklist used after mapping')\n",
    "\n",
    "args = parser.parse_args()\n",
    "FILES_DIR=args.outdir\n",
    "samples=args.samples\n",
    "\n",
    "stats_files_ext = {\n",
    "    'raw'    : 'read_count.txt',\n",
    "    'trimmed': 'trimmed.read_count.txt',\n",
    "    'mapped' : 'trimmed.bowtie.log.read_count.mapped',\n",
    "    'mapped_and_filtered': 'trimmed.masked.dedup.sorted_read_count.txt',\n",
    "    'in_broadpeaks'  : 'NA',\n",
    "    'in_narrowpeaks'  : 'trimmed.masked.dedup.sorted.bam.in_peaks.read_count.within_replicate.txt',\n",
    "    'broadPeak_count': 'trimmed.masked.dedup.sorted_peaks.broadPeak.peak_count.within_replicate.txt',\n",
    "    'narrowPeak_count': 'trimmed.masked.dedup.sorted_peaks.narrowPeak.peak_count.within_replicate.txt',\n",
    "    'NRF': 'trimmed.preseq_c_curve.percentage_unique_reads.txt',\n",
    "    'PBC': 'trimmed.PBC.txt',\n",
    "    'NSC': 'trimmed.masked.dedup.sorted.spp_cross_corr.txt'\n",
    "}\n",
    "\n",
    "if args.nonmasked:\n",
    "    stats_files_ext = {k: v.replace('.masked', '') for k,v in stats_files_ext.items()}\n",
    "\n",
    "stats_sorted = [\n",
    "    'raw',\n",
    "    'trimmed',\n",
    "    'mapped',\n",
    "    'percentage_unique',\n",
    "    'mapped_and_filtered',\n",
    "    'percentage_unique_mapped_and_filtered',\n",
    "    'in_broadpeaks',\n",
    "    'in_narrowpeaks',\n",
    "    'percent_in_broadpeaks',\n",
    "    'percent_in_narrowpeaks',\n",
    "    'broadPeak_count',\n",
    "    'narrowPeak_count',\n",
    "    'NRF',\n",
    "    'PBC',\n",
    "    'NSC',\n",
    "    'RSC'\n",
    "]\n",
    "\n",
    "print(','.join(['sample'] + stats_sorted))\n",
    "\n",
    "for s in samples:\n",
    "    stats=dd(lambda: 'N/A')\n",
    "    for k in stats_sorted:  # stats_files_ext.keys():\n",
    "        try:\n",
    "            with open('%s/%s.%s'%(FILES_DIR, s, stats_files_ext[k])) as ifile:\n",
    "                for l in ifile:\n",
    "                    vals = l.rstrip().split()\n",
    "                    if k == 'mapped':\n",
    "                        stats[k] = vals[1]\n",
    "                        try:\n",
    "                            stats['percentage_unique'] = 100*float(vals[1])/float(vals[0])\n",
    "                        except TypeError: \n",
    "                            stats['percentage_unique'] = 'N/A'\n",
    "                    elif k == 'NSC':\n",
    "                        stats['NSC']=float(vals[8])\n",
    "                        stats['RSC']=float(vals[9])\n",
    "                    elif k == 'in_narrowpeaks' or k == 'broadpeaks': \n",
    "                        stats[k] = int(vals[0])\n",
    "                        try:\n",
    "                            stats['percent_%s' % k]=100*float(stats[k])/float(stats['mapped_and_filtered'])\n",
    "                        except TypeError as ValueError:\n",
    "                            stats['percent_%s' % k]='N/A'\n",
    "                    else:\n",
    "                        try:\n",
    "                            stats[k]=int(vals[0])\n",
    "                        except:\n",
    "                            stats[k]=float(vals[0])\n",
    "        except (IOError, KeyError, ValueError):\n",
    "            pass\n",
    "    try:\n",
    "        stats['percentage_unique_mapped_and_filtered'] = 100*float(stats['mapped_and_filtered'])/stats['trimmed']\n",
    "    except ValueError:\n",
    "        stats['percentage_unique_mapped_and_filtered'] = 'N/A'\n",
    "    print(','.join([s] + [str(stats[kk]) for kk in stats_sorted]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ef43d88",
   "metadata": {},
   "source": [
    "Generate QCs for Safi_10859_250321B9_se-se-blacklist-removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0997f070",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "%%script --out blocking_job_str bash -s \"$blocking_job\"\n",
    "sbatch \\\n",
    " -p common \\\n",
    " --depend afterok:$1 \\\n",
    " /work/aeb84/processing/atac_seq/scripts/generate_qc_cell_Safi_10859_250321B9_se-se-blacklist-removal.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e42ca9b2",
   "metadata": {},
   "source": [
    "Extract blocking job id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b2b9b2",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "import re\n",
    "blocking_job = re.match('Submitted batch job (\\d+).*', blocking_job_str).group(1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3 (alex_mini3)",
   "language": "python",
   "name": "alex_mini3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
